openedx/openedx-platform
# feat: connect teams with content groups using dynamic partition generator

## Description

This PR implements the connection from the teams feature to the content groups feature [discussed](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3885662226/Connecting+Teams+to+Content+Groups) and [approved](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3885334531/Enhancements+to+Teams+to+make+learner+grouping+more+flexible?focusedCommentId=3901063179) by product. This implementation uses the dynamic partition generator extension point to associate content groups with the users that belong to a Team.

This implementation was heavily inspired by the enrollment tracks dynamic partitions.

Rejected alternatives:
- Using a fixed bridge as it is done for cohorts (pull/33105)

## Supporting information

**Product roadmap issue for this feature:**
[Enhancements to teams to make learner grouping more flexible ](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3885334531/Enhancements+to+Teams+to+make+learner+grouping+more+flexible)
[Connecting teams to content groups](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3885662226/Connecting+Teams+to+Content+Groups )

As of today, course authors can create topics or team-sets within a course. Then, associate teams to each team-set so students can join later. When connecting teams to content groups, learners can work on group assignments and access content designed by the course author specifically for that course team, as with cohorts or enrollment tracks. 

This is how this feature works behind the scenes (roughly):
- We plugged in a new `dynamic_partition_generator` function called `create_team_set_partition` that creates a `TeamUserPartition` object in memory for each course's team-set created via Pages & Resources > Teams UI in Studio.
- Each (content) group associated with each partition is dynamically generated based on the teams within the team-set when loading course content in the LMS.
- So, each time a Student loads the course blocks, the course `UserPartitionTransformer` transformer collects the blocks the current user can access. When doing so, the registered course outline processors are called. Among them is the `TeamPartitionGroupsOutlineProcessor`, which gets the information in the blocks `group_access` field and then matches it with the student's group filtering the course tree for the current user.

## Testing instructions

**Warning:** You must create a new course to avoid data inconsistencies after the test. 
Since we're adding a new entrypoint to the `setup.py` file,  you should connect to your LMS/CMS containers and run: `pip install -e .`
Then, restart them.

1. After creating your course, you'll need to turn on a course-level flag that enables content groups for teams:
Go to /admin/waffle_utils/waffleflagcourseoverridemodel
Add new flag using your new course ID:  **teams.content_groups_for_teams**

2. Enable teams for the course: Studio > Select course > Content > Pages & resources > Teams > Teams toggle
![image](https://github.com/openedx/edx-platform/assets/64440265/adff75c1-d9aa-47b2-aeb8-76d938bbc57b)

3. Create a few team sets (or groups) from the Course Authoring MFE for your course:
![image](https://github.com/openedx/edx-platform/assets/64440265/d635b2a2-464e-41ed-9e8c-a21ebdcca4c5)

4. Then, go to the LMS and create a few teams within those team-set (or topics, or groups): 
![image](https://github.com/openedx/edx-platform/assets/64440265/5a3011e1-6937-4677-96d1-8e3760afb6c7)
![image](https://github.com/openedx/edx-platform/assets/64440265/8af17988-d652-4334-9d9b-3f2b38ddf095)

5. As a student, try loading the sections of the course, you'll be able to load each course section. You'll be able to see them as usual.

6. Now, restrict topics for each team as you would with cohorts or enrollment tracks:
![image](https://github.com/openedx/edx-platform/assets/64440265/12125d5e-9c1a-44a8-81c2-af28b6e7bb05)
![image](https://github.com/openedx/edx-platform/assets/64440265/2b9e991f-ecfa-45b6-8392-d2bcf2d0ba69)
![image](https://github.com/openedx/edx-platform/assets/64440265/20f76af0-01b0-4102-87b8-5b787e29de42)

8. As a student, try loading the course sections again. You won't be able to see any content.

9. As a student, join one of the teams and try again: you'll be able to see the content within your team

10. As a student, join the other team: you'll be able to see the content within both teams

## Deadline

This effort is part of the Spanish consortium project, so it'd be ideal to merge this before the end of the project.

## Other information

**More technical info on teams:** https://openedx.atlassian.net/wiki/spaces/AC/pages/160611195
**Course author docs on teams:** https://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/course_features/teams/teams_setup.html#teams-overview

### PR Sandbox

**Settings**

```yaml
TUTOR_GROVE_WAFFLE_FLAGS:
- name: teams.content_groups_for_teams
  everyone: true
- name: teams.enable_teams_app
  everyone: true
```
